sdk/src/main/java/com/hedera/hashgraph/sdk/Network.java

```diff
@@ -158,6 +160,44 @@ private Network setLedgerIdInternal(@Nullable LedgerId ledgerId, @Nullable Map<A
         return this;
     }
 
+    void setAddressBook(NodeAddressBook addressBook) {
+        Map<AccountId, NodeAddress> newAddressBook = addressBook.getNodeAddresses().stream()
+            .filter(nodeAddress -> Objects.nonNull(nodeAddress.getAccountId()))
+            .collect(Collectors.toMap(NodeAddress::getAccountId, Function.identity(),
+                /*
+                * Here we index by AccountId ignoring any subsequent entries with the same AccountId.
+                *
+                * Currently, this seems to be needed when reloading predefined address book for testnet which contains
```

Comment:
It's weird that there are multiple entries with the same AccountId, can you give more information why this is happening if you found something during your testing?

Reply:
What do you think about simplifying this logic:

```java
 void setAddressBook(NodeAddressBook addressBook) {
        var newNodeAddresses = addressBook.getNodeAddresses();

        if (this.addressBook != null) {
            for (NodeAddress newNodeAddress : newNodeAddresses) {
                NodeAddress previousNodeAddress = this.addressBook.get(newNodeAddress.getAccountId());

                if (previousNodeAddress != null) {
                    ByteString certHash = newNodeAddress.getCertHash();
                    if (certHash == null || certHash.isEmpty()) {
                        newNodeAddress.setCertHash(previousNodeAddress.getCertHash());
                    }
                }
                this.addressBook.put(newNodeAddress.getAccountId(), newNodeAddress);
            }
        } else {
            this.addressBook = new HashMap<>();
            for (NodeAddress newNodeAddress : newNodeAddresses) {
                this.addressBook.put(newNodeAddress.getAccountId(), newNodeAddress);
            }
        }

        for (var node : nodes) {
            node.setAddressBookEntry(this.addressBook.get(node.getAccountId()));
        }
    }
```

Would this work in all cases (unit test is passing)?

Reply:
> It's weird that there are multiple entries with the same AccountId, can you give more information why this is happening if you found something during your testing?

@0xivanov I believe I saw this in the predefined address book for testnet, I'm not sure why.

Reply:
> What do you think about simplifying this logic like this:

It looks pretty similar to what I had, the only real significant difference I noticed just visually comparing is that this version doesn't seem to remove entries in the addressbook in the case when the account-id is missing from the latest addressbook.

By the way, as far as I can tell aside from this method which updates `Network::addressbook` then loops through it to update nodes, `Network::addressbook` appears to be unused, except for two tests. You might be able to remove `Network::addressbook` in that case, if we still want to try and preserve old TLS parameters, you can check the address-book-entry in the node before overwriting it.

Reply:
> the only real significant difference I noticed just visually comparing is that this version doesn't seem to remove entries in the addressbook in the case when the account-id is missing from the latest addressbook.

What piece of logic are you referring to?

Reply:
> What piece of logic are you referring to?

This version only doesn't seem to remove entries from `this.addressBook` in the case when the account-id is missing from the latest `addressBook`.

```
 void setAddressBook(NodeAddressBook addressBook) {
        var newNodeAddresses = addressBook.getNodeAddresses();

        if (this.addressBook != null) {
            for (NodeAddress newNodeAddress : newNodeAddresses) {
                // ...
                this.addressBook.put(newNodeAddress.getAccountId(), newNodeAddress);
            }
        }
        // ...
    }
```

Compared with this version:

```
    void setAddressBook(NodeAddressBook addressBook) {
        Map<AccountId, NodeAddress> newAddressBook = addressBook.getNodeAddresses().stream()
            .filter(nodeAddress -> Objects.nonNull(nodeAddress.getAccountId()))
            .collect(Collectors.toMap(NodeAddress::getAccountId, Function.identity(),
                (a, b) -> a));
        // ...
        this.addressBook = newAddressBook;
    }
```

Reply:
I see what you mean,
I wrote a unit test for this flow and even `network.setNetwork()` does not handle the scenario when the accountID is null. It populates the nodes list with this invalid accountID and the test fails.
I am not sure if it is even possible to have a node without account id and if it is necessary to handle this case.

Reply:
Ok, will keep your implementation.